tritium-ids/src/main/java/com/palantir/tritium/ids/UniqueIds.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.ids;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+
+public final class UniqueIds {
+    private static final SafeLogger log = SafeLoggerFactory.get(UniqueIds.class);
+
+    private UniqueIds() {}
+
+    private static final ThreadLocal<Random> random = ThreadLocal.withInitial(() -> {
+        try {
+            return SecureRandom.getInstance("SHA1PRNG");
+        } catch (NoSuchAlgorithmException e) {
+            log.warn("Falling back to default SecureRandom", e);
+            return new SecureRandom();
+        }
+    });
+
+    /**
+     * Returns a unique {@link UUID} using.
+     */
+    public static UUID v4PseudoRandomUuid() {
+        return v4RandomUuid(ThreadLocalRandom.current());
+    }
+
+    /**
+     * Returns a unique {@link UUID}.
+     */
+    public static UUID v4RandomUuid() {
+        return v4RandomUuid(random.get());
+    }
+
+    @VisibleForTesting
+    static UUID v4RandomUuid(Random rand) {
+        return v4RandomUuid(bytes(rand));
+    }
+
+    private static UUID v4RandomUuid(byte[] data) {
+        return ietfUuid(data, 0x40);
+    }
+
+    private static UUID ietfUuid(byte[] data, int version) {
+        Preconditions.checkArgument(data.length == 16, "Invalid data length, expected 16 bytes");
+        data[6] = (byte) ((data[6] & 0x0f) | version); // version 4
+        data[8] = (byte) ((data[8] & 0x3f) | 0x80); // IETF variant
+
+        long mostSigBits = 0;
+        for (int i = 0; i < 8; i++) {
+            mostSigBits = (mostSigBits << 8) | (data[i] & 0xff);
+        }
+
+        long leastSigBits = 0;
+        for (int i = 8; i < 16; i++) {
+            leastSigBits = (leastSigBits << 8) | (data[i] & 0xff);
+        }
+
+        return new UUID(mostSigBits, leastSigBits);
+    }
+
+    private static byte[] bytes(Random rand) {
+        byte[] data = new byte[16];
+        rand.nextBytes(data);
+        return data;
+    }
```

Comment:
I wonder if it's reasonable to use a pair of `random.nextLong()` results?

Reply:
I can benchmark that.

My sense is that the contention is actually on the SecureRandom instance method invocations and not the underlying entropy source, so the two calls would contend more for the cases where there is a single shared `SecureRandom`.

I will also try out per processor pool of random bytes.

Reply:
I tested out a few iterations of pooled byte arrays in [ds/ids-pool-chunks](https://github.com/palantir/tritium/compare/ds/ids-pool-chunks?expand=1) and tweaking the `CHUNKS_TO_ALLOCATE`, but `ThreadLocal<SecureRandom>` for `v4RandomUuid` seems worth it for near linear core scaling in `v4RandomUuidMax`, compared to the negative scaling due to contention in `jdkRandomUuidMax`.



### `ThreadLocal<SecureRandom>` (this branch)

```
16x Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz cores
VM version: JDK 17.0.8, OpenJDK 64-Bit Server VM, 17.0.8+7-LTS
Benchmark                          Mode  Cnt   Score   Error   Units
UuidBenchmark.jdkRandomUuidMax    thrpt    3   1.689 ? 1.397  ops/us
UuidBenchmark.jdkRandomUuidOne    thrpt    3   2.488 ? 0.252  ops/us
UuidBenchmark.jugSecureRandomMax  thrpt    3   2.044 ? 1.434  ops/us
UuidBenchmark.secureRandomMax     thrpt    3   2.053 ? 1.473  ops/us
UuidBenchmark.threadLocalMax      thrpt    3  28.340 ? 2.180  ops/us
UuidBenchmark.twoLongsMax         thrpt    3   0.699 ? 1.036  ops/us
UuidBenchmark.v4RandomUuidMax     thrpt    3  27.893 ? 2.183  ops/us
UuidBenchmark.v4RandomUuidOne     thrpt    3   3.393 ? 0.312  ops/us
```

```
MacBook Pro 2021 10x aarch64 cores
VM version: JDK 17.0.8.1, OpenJDK 64-Bit Server VM, 17.0.8.1+8-LTS
Benchmark                          Mode  Cnt   Score   Error   Units
UuidBenchmark.jdkRandomUuidMax    thrpt    3   2.265 ? 0.233  ops/us
UuidBenchmark.jdkRandomUuidOne    thrpt    3   3.864 ? 0.064  ops/us
UuidBenchmark.jugSecureRandomMax  thrpt    3   2.886 ? 0.186  ops/us
UuidBenchmark.secureRandomMax     thrpt    3   2.847 ? 0.851  ops/us
UuidBenchmark.threadLocalMax      thrpt    3  26.719 ? 5.500  ops/us
UuidBenchmark.twoLongsMax         thrpt    3   1.125 ? 0.193  ops/us
UuidBenchmark.v4RandomUuidMax     thrpt    3  36.466 ? 4.738  ops/us
UuidBenchmark.v4RandomUuidOne     thrpt    3   4.546 ? 0.461  ops/us
```

### [`LinkedBlockingDeque<byte[]>`](https://github.com/palantir/tritium/compare/ds/ids-pool-chunks?expand=1)  , `ConcurrentLinkedDeque<byte[]>` & `synchronizedDeque(new ArrayDeque(CHUNKS_TO_ALLOCATE)` is similar

```
16x Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz cores
VM version: JDK 17.0.8, OpenJDK 64-Bit Server VM, 17.0.8+7-LTS
Benchmark                          Mode  Cnt  Score   Error   Units
UuidBenchmark.jdkRandomUuidMax    thrpt    3  1.680 ? 1.495  ops/us
UuidBenchmark.jdkRandomUuidOne    thrpt    3  2.500 ? 0.073  ops/us
UuidBenchmark.jugSecureRandomMax  thrpt    3  2.003 ? 1.428  ops/us
UuidBenchmark.secureRandomMax     thrpt    3  3.678 ? 0.135  ops/us
UuidBenchmark.threadLocalMax      thrpt    3  2.143 ? 0.564  ops/us
UuidBenchmark.twoLongsMax         thrpt    3  0.723 ? 1.223  ops/us
UuidBenchmark.v4RandomUuidMax     thrpt    3  3.531 ? 0.526  ops/us
UuidBenchmark.v4RandomUuidOne     thrpt    3  2.935 ? 0.135  ops/us
```

```
MacBook Pro 2021 10x aarch64 cores
VM version: JDK 17.0.8.1, OpenJDK 64-Bit Server VM, 17.0.8.1+8-LTS
Benchmark                          Mode  Cnt   Score   Error   Units
UuidBenchmark.jdkRandomUuidMax    thrpt    3   2.305 ? 0.251  ops/us
UuidBenchmark.jdkRandomUuidOne    thrpt    3   3.843 ? 0.098  ops/us
UuidBenchmark.jugSecureRandomMax  thrpt    3   2.952 ? 0.177  ops/us
UuidBenchmark.secureRandomMax     thrpt    3   5.403 ? 0.117  ops/us
UuidBenchmark.threadLocalMax      thrpt    3  10.920 ? 8.761  ops/us
UuidBenchmark.twoLongsMax         thrpt    3   1.153 ? 0.128  ops/us
UuidBenchmark.v4RandomUuidMax     thrpt    3   5.402 ? 0.284  ops/us
UuidBenchmark.v4RandomUuidOne     thrpt    3   4.824 ? 0.228  ops/us
```